lib/proxy/clientinterceptor.go

```diff
@@ -0,0 +1,73 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package proxy
+
+import (
+	"context"
+
+	"google.golang.org/grpc"
+)
+
+// streamClientInterceptor is GPRC stream interceptor that handles:
+// - reporting / metrics.
+// - converting errors to the appropriate grpc status code.
+func streamClientInterceptor(metrics *clientMetrics) grpc.StreamClientInterceptor {
```

Comment:
Did you consider using a `stats.Handler`? 

https://pkg.go.dev/google.golang.org/grpc/stats#Handler

Reply:
The interceptor was initially added to capture and convert errors to grpc type errors. I then thought to add the metrics in the interceptor itself since the infrastructure is already there. But you're right stats.Handler is looks cleaner and is easier to follow. If you have any strong opinions on it, I would move the metrics out of the interceptor, it shouldn't be too much work

Reply:
I don't necessarily have a strong opinion about it. While looking at some of the code added to handle metrics I wondered if some of it could be eliminated/simplified by using a `stats.Handler`. Most implementations tracking grpc metrics that I have seen have used a `stats.Handler`, cockroachdb for example: https://github.com/cockroachdb/cockroach/blob/4d987b3f0788c98fa358715b336a1c55e254ee50/pkg/rpc/stats_handler.go